# EE 467 Homework 1: Twitter Spam Detection

Welcome to the first homework of EE 467! Here we will apply the ML pipline introduced in the second lab to **Twitter spam message detection problems**. In the past few years, social medias like Twitter has become the target of spam posting. Automatic spam posts can not only affect user experience, but they can also spread misinformation and subtly affect public opinions.

In this homework, we will work on a mini Twitter message dataset extracted from the CRESCI-2017 dataset. This dataset contains 50000 Twitter messages, half of which are sent by real Twitter users and half of which generated by spam bots. Your task is to train classifiers that can tell these two kinds of messages apart. This time, you'll go deeper and try multiple feature extractors and classifiers. Like the labs, you will need to fill out blanks in code cells. You are also going to answer a few questions and **optionally** work on improvements to existing pipeline completely on your own.

## Before You Start

1. You can discuss on Slack if you have questions and want to seek help; however, please try your best to **limit the scope of your question** and **avoid asking directly for answers**. You should also **avoid copy-pasting answers and code** from others.
2. You **may use AI assistants** (e.g., ChatGPT, Gemini, etc.) to support your work. If you do, you must **clearly acknowledge** this by stating:

- **The name of the AI assistant**, and  
- **How it was used** (e.g., brainstorming, clarifying concepts, debugging, editing, generating examples)

You **must not copy and paste** AI-generated content directly into your submission without **fully understanding** what it does and being able to **explain it in your own words**.

3. You can **optionally work on this homework with one other student of this course as a team**, but **each of you needs to submit the homework individually**. If you choose to form a group, please include the name of your teammate here: _(Name of your teammate)_.

## Feature Extraction

We will begin our ML pipeline with feature extraction in this homework, as the dataset is good enough and does not need any special pre-processing. As usual, we will load the dataset into memory, extract samples features into vectors and split the dataset into training and test sets.

In [1]:
import pandas as pd

## [ TODO ]
# 1) Load all data from `twitter-mini.csv` into variable `data`
# 2) Preview first few samples
data = pd.read_csv("twitter-mini.csv")
data.head()

text  label
0  @baydiangirl it definitely moved some church f...      0
1  @suebob Looks like #TXlege made up new rules f...      0
2  RT @saleemchat: #WJChat #A1B 5 yrs: People ski...      0
3                            @SajidaBalouch Huhhhhh       0
4              @Dukester_94 im the lad of all bibles      0

In lab 2 we have learned how to use the `CountVectorizer` feature extractor, which counts the occurance of each token (letter here) in a sample (Twitter message text here). This time, we will also try `TfidfVectorizer` and see which feature extractor performs better. Here we will introduce the algorithm it implemented, **TF-IDF**:

> **TF-IDF** is the abbreviation for **term frequency–inverse document frequency**, and is defined as the product of the two frequency. The most classical version of TF-IDF is:
>
> $$
tf-idf(t) = tf(t, d) \cdot idf(t, D) = \frac{f_{t,d}}{|d|} \log \frac{|D|}{|D_t|}
$$
>
> Where $t$ is the term, $D$ is the collection of all documents, $d$ is one of the document that contains $t$ and $D_t$ is the subset of documents that contains term $t$. The advantage of TF-IDF is that frequent and useless words, such as "the", "a" and "an" will receive lower weight, since they appear in almost all documents. On the other hand, unique yet frequent (in any $d \in D_t$) words will get higher weight. Research has shown that TF-IDF is easy to implement and performs relatively well.

Before we run these feature extractors on our dataset, we need to convert each message into a sequence of terms. Since Twitter messages often contain emojis, abbreviations, hashtags and other composition of characters, we can't use the word-level analyzer as we have done in lab 2. Instead, we will generate **character N-grams** from these messages.

> A **character N-gram** consists of N consecutive characters. Here we show all character 1, 2, 3 and 4-grams for word "cold":
> <img src="attachment:char-ngram.jpg" width="50%" />

Both `CountVectorizer` and `TfidfVectorizer` provide support for character N-grams. Refer to `sklearn`'s documents to build a character N-gram analyzer.

In [8]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

## [ TODO ]
# 1) Extract features of all messages into `feat1` with `CountVectorizer` using CHARACTER trigrams
# 2) Repeat with `TfidfVectorizer` and save generated features as `feat2`
vec1 = CountVectorizer(analyzer='char', ngram_range=(3, 3))
feat1 = vec1.fit_transform(data["text"])

vec2 = TfidfVectorizer(analyzer='char', ngram_range=(3, 3))
feat2 = vec2.fit_transform(data["text"])


We will also try combining two kinds of feature generated above into a single feature. We will use `scipy.sparse.hstack` to concatenate corresonding features for each sample. Remember, concatenated features matrix should have same amount of samples as `feat1` and `feat2`.

In [9]:
from scipy.sparse import hstack

## [ TODO ]
# Concatenate token count features and TF-IDF features on the feature dimension,
# name the combined feature matrix `feat_comb`.
feat_comb = hstack([feat1, feat2])

Finally, let's take a look at the labels. In our dataset, 0 represents messages from real Twitter users and 1 represents spam messages.

In [10]:
y = data["label"].values

## Training

As usual, we will start by doing a 80% / 20% training-test split. Like lab 2, we suggest you to **use the same random seed for all splits** for comparable and reproducible results.

In [11]:
from sklearn.model_selection import train_test_split

## [ TODO ]
# 1) Split token count features into 80/20 training and test sets
# 2) Split TF-IDF features into 80/20 training and test sets
# 3) Split combined features into 80/20 training and test sets

RANDOM_SEED = 42

# Token count features
X1_train, X1_test, y1_train, y1_test = train_test_split(feat1, y, test_size=0.2, random_state=RANDOM_SEED)

# TF-IDF features
X2_train, X2_test, y2_train, y2_test = train_test_split(feat2, y, test_size=0.2, random_state=RANDOM_SEED)

# Combined features
X_comb_train, X_comb_test, y_comb_train, y_comb_test = train_test_split(feat_comb, y, test_size=0.2, random_state=RANDOM_SEED)

We are now ready to train **binary classifiers** for spam message detection. Apart from the `LogisticRegression` classifier you have already seen and used during Lab 1, we will try a few more classifiers:

* `MultinomialNB` (Naive Bayes):  
  This classifier is based on the **Naive Bayes** assumption that, *given the class label*, individual features are conditionally independent.  
  For text classification, `MultinomialNB` is used with **token count** or **TF-IDF-like** features.

  It models the probability of a message belonging to a class $ y \in \{\text{spam}, \text{ham}\} $ as:

  $$
  P(y \mid \mathbf{x}) \propto P(y)\prod_{j=1}^{d} P(x_j \mid y)
   $$

  where $\mathbf{x} = (x_1, x_2, \dots, x_d)$ is the feature vector (e.g., word counts), and $P(x_j \mid y)$ is learned from training data.  
  The predicted label is chosen using:

   $$
  y_{pred} = \arg\max_y \; P(y \mid \mathbf{x})
   $$

  In practice, the computation is done in the **log domain** to avoid numerical underflow:

  $$
  y_{pred} = \arg\max_y \left( \log P(y) + \sum_{j=1}^{d} x_j \log P(x_j \mid y) \right)
  $$

* `LinearSVC`:  
  This classifier is based on **linear SVM**, a popular algorithm for binary classification. The linear SVM algorithm treats the feature vector of each sample as a point in a linear space, and tries to find a hyperplane that separates points of the two classes as well as possible.  

  In practice, we optimize a **squared hinge loss with regularization**, since the features may not be perfectly linearly separable:

   $$
  l_{hinge} = \max(0, y(1 - \mathbf{x}^T \mathbf{w}))^2 + \beta ||\mathbf{w}||_2^2
   $$

  Similarly, the classification result is determined by:

  $$
  y_{pred} = sign(1 - \mathbf{x}^T \mathbf{w})
 $$

We have three kinds of features (**token count**, **TF-IDF**, and **combined**) and three kinds of classifiers. Here we will only try three combinations and compare their performance:

* `MultinomialNB` on token count features  
* `LogisticRegression` on TF-IDF features  
  - Try increasing the number of iterations if a warning of non-convergence pops up  
* `LinearSVC` on combined features  
  - For linear SVM, you can ignore the non-convergence warning as it does not affect performance  


In [12]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn.svm import LinearSVC

## [ TODO ]
# 1) Train a `MultinomialNB` classifier called `model1` on token count training set
# 2) Train a `LogisticRegression` classifier called `model2` on TF-IDF training set
# 3) Train a `LinearSVC` classifier called `model_comb` on combined features training set

model1 = MultinomialNB()
model1.fit(X1_train, y1_train)

model2 = LogisticRegression(max_iter=1000)
model2.fit(X2_train, y2_train)

model_comb = LinearSVC()
model_comb.fit(X_comb_train, y_comb_train)

c:\Users\linwh\anaconda3\Lib\site-packages\sklearn\svm\_base.py:1249: ConvergenceWarning: Liblinear failed to converge, increase the number of iterations.
  warnings.warn(


LinearSVC()

## Evaluation

Finally, it's time to evaluate our models. You probably noticed that we have repeated similar code for different feature extraction and classification models. To avoid that, we will use `for` loops to make evaluation simpler:

In [13]:
from sklearn.metrics import precision_recall_fscore_support, confusion_matrix, accuracy_score

features = [("Token count", feat1), ("TF-IDF", feat2), ("Combined", feat_comb)]
models = [("Naive Bayes", model1), ("Logistic", model2), ("SVM", model_comb)]

## [ TODO ]
# Complete missing code in the loop body
for i, ((feat_name, feat), (model_name, model)) in enumerate(zip(features, models)):
    # Begin of metrics for combination
    print("====================")
    print(f"{i+1}. {model_name} on {feat_name}:\n")

    # 1) Split features into training and test set
    # (Note: you should use the same random number seed as the one in the training-test split above)
    X_train, X_test, y_train, y_test = train_test_split(feat, y, test_size=0.2, random_state=RANDOM_SEED)

    # 2) Predict labels for training data
    pred_train = model.predict(X_train)

    # 3) Print metrics for training set (precision, recall and F1 metrics)
    precision_train, recall_train, f1_train, _ = precision_recall_fscore_support(y_train, pred_train, average='binary')
    print("Training metrics:", f"precision={precision_train:.4f}, recall={recall_train:.4f}, f1={f1_train:.4f}")
    # Confusion matrix
    print("Training confusion matrix:\n", confusion_matrix(y_train, pred_train), "\n")
    # Accuracy
    print("Train accuracy:", accuracy_score(y_train, pred_train), "\n")

    # 4) Predict labels for test data
    pred_test = model.predict(X_test)

    # 5) Print metrics for test set (precision, recall and F1 metrics)
    precision_test, recall_test, f1_test, _ = precision_recall_fscore_support(y_test, pred_test, average='binary')
    print("Test metrics:", f"precision={precision_test:.4f}, recall={recall_test:.4f}, f1={f1_test:.4f}")
    # Confusion matrix
    print("Test Confusion Matrix:\n", confusion_matrix(y_test, pred_test), "\n")
    # Accuracy
    print("Test Accuracy:", accuracy_score(y_test, pred_test))

    # End of metrics for combination
    print("====================\n")

1. Naive Bayes on Token count:

Training metrics: precision=0.9847, recall=0.9127, f1=0.9473
Training confusion matrix:
 [[19694   284]
 [ 1748 18274]] 

Train accuracy: 0.9492 

Test metrics: precision=0.9820, recall=0.9084, f1=0.9438
Test Confusion Matrix:
 [[4939   83]
 [ 456 4522]] 

Test Accuracy: 0.9461

2. Logistic on TF-IDF:

Training metrics: precision=0.9918, recall=0.9387, f1=0.9645
Training confusion matrix:
 [[19822   156]
 [ 1228 18794]] 

Train accuracy: 0.9654 

Test metrics: precision=0.9861, recall=0.9265, f1=0.9554
Test Confusion Matrix:
 [[4957   65]
 [ 366 4612]] 

Test Accuracy: 0.9569

3. SVM on Combined:

Training metrics: precision=0.9999, recall=0.9986, f1=0.9993
Training confusion matrix:
 [[19977     1]
 [   28 19994]] 

Train accuracy: 0.999275 

Test metrics: precision=0.9487, recall=0.9482, f1=0.9485
Test Confusion Matrix:
 [[4767  255]
 [ 258 4720]] 

Test Accuracy: 0.9487



# Questions
Please answer the following questions. You can just write down the answer below each question in this cell.


1. Please compute `tf-idf` of word-level **bigrams** for the following three messages **by hand**. You should use the TF-IDF definition introduced in class.

  * Message 1: "vegetables are good for health"
  * Message 2: "fruits are good source of vitamins"
  * Message 3: "proteins are good for health"

  Compute and sort intermediate and final results for **all occurring bigrams in the following order**:
  
  ```
  ["vegetables are", "are good", "good for", "for health", "fruits are", "good source", "source of", "of vitamins", "proteins are"]
  ```
  
  Then complete the following questions:

  1. Term Frequency (`tf`) of **bigrams for message 2**.
  2. Inverse Document Frequency (`idf`) of **bigrams for message 2**.
  3. `tf-idf` of **bigrams for message 2**.
  4. Is the `tf-idf` value for the term "are good" equal to 0? If it's 0, can you explain why it makes sense to have 0 for that term?
  5. If we consider unigram (1-gram) based bag of words, how many words in these three messages would be assigned a 0 value tf-idf? And please explain why.

---

**Answers for Question 1:**

**Document analysis:**
- Message 1 bigrams: vegetables are, are good, good for, for health
- Message 2 bigrams: fruits are, are good, good source, source of, of vitamins
- Message 3 bigrams: proteins are, are good, good for, for health

**Document frequency (|D_t|):** vegetables are(1), are good(3), good for(2), for health(2), fruits are(1), good source(1), source of(1), of vitamins(1), proteins are(1). |D| = 3.

**1.1 Term Frequency (tf) of bigrams for message 2:**  
Using tf(t,d) = f_{t,d} / |d|, where |d| = 5 (number of bigrams in message 2). Bigrams not in message 2 have tf = 0. Listed in the required order (all 9 bigrams):
- vegetables are: tf = 0
- are good: tf = 1/5 = 0.2
- good for: tf = 0
- for health: tf = 0
- fruits are: tf = 1/5 = 0.2
- good source: tf = 1/5 = 0.2
- source of: tf = 1/5 = 0.2
- of vitamins: tf = 1/5 = 0.2
- proteins are: tf = 0

**1.2 Inverse Document Frequency (idf) of bigrams for message 2:**  
Using idf(t) = log(|D|/|D_t|). Listed in the required order (all 9 bigrams):
- vegetables are: idf = log(3/1) = log(3)
- are good: idf = log(3/3) = 0
- good for: idf = log(3/2)
- for health: idf = log(3/2)
- fruits are: idf = log(3/1) = log(3)
- good source: idf = log(3/1) = log(3)
- source of: idf = log(3/1) = log(3)
- of vitamins: idf = log(3/1) = log(3)
- proteins are: idf = log(3/1) = log(3)

**1.3 tf-idf of bigrams for message 2:**  
Listed in the required order (all 9 bigrams):
- vegetables are: 0 × log(3) = 0
- are good: 0.2 × 0 = 0
- good for: 0 × log(3/2) = 0
- for health: 0 × log(3/2) = 0
- fruits are: 0.2 × log(3)
- good source: 0.2 × log(3)
- source of: 0.2 × log(3)
- of vitamins: 0.2 × log(3)
- proteins are: 0 × log(3) = 0

**1.4** Yes, the tf-idf value for "are good" is 0. It appears in all three documents, so idf = log(1) = 0. This is reasonable because "are good" does not help distinguish between documents; it carries no discriminative information.

**1.5** For unigrams, the unique words are: vegetables, are, good, for, health, fruits, source, of, vitamins, proteins. The words "are" and "good" appear in all three documents, so their idf = log(1) = 0 and thus tf-idf = 0. **2 words** would have 0 tf-idf. Words that appear in every document receive zero weight because they do not help separate documents.

2. What does `support` mean in the `classification_report`?

**Answer:** In `classification_report`, **support** is the number of true (actual) samples for each class in the dataset being evaluated. For example, in our test set with 10,000 samples, if there are 5,000 ham (label 0) and 5,000 spam (label 1), the support for each class would be 5000. Support indicates how many ground-truth examples of each class were used to compute the precision, recall, and F1-score for that class.

3. If the training data is highly unbalanced, is `accuracy` a good metric for the classifiers' performance? Does a high accuracy necessarily means a high f1-score? Compare `accuracy` with `f1-score`. Based on your criterion, which type of classifier is better? Please give some justifications to support your answer.

**Answer:** When the training data is highly unbalanced, **accuracy is generally not a good metric**. For example, if 99% of samples are negative, a classifier that always predicts negative would achieve 99% accuracy but would miss all positive samples, with F1 = 0 for the positive class. **High accuracy does not imply high F1-score**; accuracy can be inflated by correctly predicting the majority class while failing on the minority class.

**Accuracy vs F1-score:** Accuracy is (TP+TN)/(TP+TN+FP+FN), treating all classes equally. F1-score balances precision and recall for the positive class and is more informative when the positive class is rare or when false negatives and false positives are both important.

Using F1-score as the criterion (dataset is balanced: 50% ham, 50% spam):
- Naive Bayes on Token count: Test F1 = 0.9438, Test Accuracy = 0.9461
- Logistic on TF-IDF: Test F1 = 0.9554, Test Accuracy = 0.9569
- SVM on Combined: Test F1 = 0.9486, Test Accuracy = 0.9488

**The Logistic Regression on TF-IDF features is the best classifier**, with the highest test F1 (0.9554) and test accuracy (0.9569). It also has the best precision (0.9861) and recall (0.9265) on the test set, with fewer misclassified samples (431 total errors) compared to Naive Bayes (539) and SVM (512).

4. For most cases, the amount of bad data is much less than good data. How can you mitigate the data imbalance issue?

**Answer:** Several approaches can mitigate data imbalance:

1. **Resampling:**
   - **Oversampling:** Duplicate or synthesize minority-class samples (e.g., SMOTE) to balance the dataset.
   - **Undersampling:** Reduce the majority class by randomly removing samples until classes are more balanced.

2. **Class weights:** Assign higher loss weights to the minority class so the model pays more attention to it (e.g., `class_weight='balanced'` in scikit-learn).

3. **Collect more data:** Acquire additional minority-class samples to improve representation.

4. **Evaluation metrics:** Use metrics suited to imbalanced data (e.g., F1-score, precision-recall curve, AUC-PR) instead of accuracy.

5. **Threshold adjustment:** Lower the decision threshold for the positive class to increase recall at the cost of precision, or tune the threshold using validation data.

6. **Ensemble methods:** Use techniques like EasyEnsemble or BalancedBaggingClassifier that combine multiple undersampled subsets with the minority class.

## Model Improvements (Optional; Not Included in Homework Score)

As we have pointed out in the first class, machine learning pipeline does not end with a single iteration of training, testing and evaluation. Here, you will repeat these process with different hyper-parameters and try to improve the spam detection model so that it performs even better.

Below are two **potential improvements** that may be beneficial to the performance. You are encouraged to complete either (or both) of them. Your response to these questions are optional and won't be graded, but we will provide feedbacks to your code and analysis if you choose to answer them. You can add code and text cells below the questions as needed.

1. We use character-level trigrams as tokens during the feature extraction stage, however $N = 3$ is a rather arbitrary choice. What will the performance be like if we use shorter N-grams, such as $N = 1$ and $N = 2$? What about longer N-grams, such as $N = 4$? What if we combine different N-grams, like $N = 1, 2, 3$? Modify the feature extraction code above and play with different (range of) $N$ values. What are **the best (range of) $N$ value on the test set of combined features and SVM classifier**? What metrics did you use to reach this conclusion and why do you think this (range of) $N$ value is the best?

2. During the above training and evaluation process, we are only evaluating three combination of feature extractors and classifiers. Can you try to rewrite the training and evaluation code, such that **every combination** of feature extractors and classifiers can be evaluated? What metrics did you use to reach this conclusion and what is the best combination of feature extractor and classifier in terms of test set performance?

## References

1. CRESCI-2017 Dataset: https://botometer.osome.iu.edu/bot-repository/datasets.html
2. TF-IDF: https://en.wikipedia.org/wiki/Tf%E2%80%93idf
3. N-grams: https://en.wikipedia.org/wiki/N-gram
4. `scikit-learn` documentation: https://scikit-learn.org/stable/modules/classes.html
5. Ridge regression: https://en.wikipedia.org/wiki/Tikhonov_regularization
6. SVM: https://en.wikipedia.org/wiki/Support-vector_machine
7. Lasso: https://en.wikipedia.org/wiki/Lasso_(statistics)
8. ElasticNet: https://en.wikipedia.org/wiki/Elastic_net_regularization
9. Kernel method: https://en.wikipedia.org/wiki/Kernel_method